File: LeetCode/面试经典 150 题/数组_字符串/238. 除自身以外数组的乘积/238.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans;
        ans.resize(n);
        int last = 1;
        for (int i = n - 1; i >= 0; i--) {
            last *= nums[i];
            ans[i] = last;
        }
        last = 1;
        for (int i = 0; i < n; i++) {
            ans[i] = last * (i < n - 1 ? ans[i + 1] : 1);
            last *= nums[i];
        }
        return ans;
    }
};

int main() {
    vector<int> x = {1,2,3,4};
    Solution s;
    auto ans = s.productExceptSelf(x);
    for (auto x : ans) { printf("%d ", x); }
    return 0;
}

File: LeetCode/LeetCode 75/数组 字符串/238. 除自身以外数组的乘积/238.java
// prefix & suffix array
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] d = nums;
        int[] pre = new int[d.length];
        int[] suf = new int[d.length];
        // init
        pre[0] = d[0];
        for (int i = 1; i < pre.length; i++) {
            pre[i] = pre[i - 1] * d[i];
        }
        suf[suf.length - 1] = d[suf.length - 1];
        for (int i = suf.length - 2; i >= 0; i--) {
            suf[i] = suf[i + 1] * d[i];
        }
        // ans
        int[] ans = new int[d.length];
        for (int i = 0; i < d.length; i++) {
            ans[i] = (i > 0 ? pre[i - 1] : 1)
                * ((i < d.length - 1) ? suf[i + 1] : 1);
        }
        return ans;
    }
}